Authority doc (#473)
spatialos/UnrealGDK: SpatialGDK/Documentation/content/authority.md

# Authority in the GDK
To work with authority in the GDK, it’s useful to refresh on authority in Unreal’s native networking.

## Unreal networking authority

> In native-Unreal networking, the single server has absolute authority over all replicated Actors,  while clients have only proxies of Actors. Unreal models this using the `Role` and `RemoteRole` fields within an Actor.

Reviewer: ```suggestion
In native-Unreal networking, the single server has absolute authority over all replicated Actors, while clients have only proxies of Actors. Unreal models this using the `Role` and `RemoteRole` fields within an Actor.
```

> For the majority of use-cases, these fields have the values:

Reviewer 2: use-cases > use cases

Actor on **server**:

* `Role = ROLE_Authority`
* `RemoteRole = ROLE_SimulatedProxy`

Actor on **client**:

* `Role = ROLE_SimulatedProxy`
* `RemoteRole = ROLE_Authority`

Actors that have an [owning connection (Unreal documentation)](https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/OwningConnections) are slightly different. An Actor can have an owning connection if:

* It is a `PlayerController` with an associated `NetConnection`.

> * It is a `Pawn` which is possessed by a `PlayerController` that has an associated `NetConnection`

Reviewer 3: ```suggestion
* It is a `Pawn` which is possessed by a `PlayerController` that has an associated `NetConnection`.
```

* Its `Owner` is set to an Actor that has an owning connection.

> For example, if a client has a `PlayerController` which possesses a `Character` which is holding a gun, all three (the gun, `Character` and `PlayerController`) have an owning connection.

Reviewer 4: which is holding a gun, and the gun's `Owner` is set to `Character`

Reviewer 4: I know it's implied, but I think it's better to be explicit here.

In any of these cases, Unreal authority looks like:

Actor on **server** with an **owning connection**:

* `Role = ROLE_Authority`
* `RemoteRole = ROLE_AutonomousProxy`

Actor on **owning client**:

* `Role = ROLE_AutonomousProxy`
* `RemoteRole = ROLE_Authority`

Actor on **non-owning client**:

* `Role = ROLE_SimulatedProxy`
* `RemoteRole = ROLE_Authority`

> `ROLE_AutonomousProxy` indicates that this Actor is owned by a client and can receive client RPCs called on the Actor, as well as trigger server RPCs which are handled by the server. For more information on owning connection see the [Unreal doccumentation](https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/OwningConnections)

Reviewer 3: ```suggestion
`ROLE_AutonomousProxy` indicates that this Actor is owned by a client and can receive client RPCs called on the Actor, as well as trigger server RPCs which are handled by the server. For more information on owning connection see the [Unreal doccumentation](https://docs.unrealengine.com/en-us/Gameplay/Networking/Actors/OwningConnections).
```

Reviewer 5: How come we define this below where it's used. What about putting a description of the various Roles above and then showing how they are used on clients and servers.

Author: Good point

> For more information about Actor roles, see [Unreal’s Actor Role documentation](https://wiki.unrealengine.com/Replication#A_Guide_To_Network_Roles).

Reviewer: Unreal calls this `network roles`.

## GDK authority

> As the GDK works with multiple [server-workers]({{urlRoot}}/content/glossary#workers), rather than a single server, authority needs to be dictated by SpatialOS so that authority is shared between server-workers. This means server-workers have authority over some Actors but don’t have authority over other Actors, depending on how SpatialOS assigns authority. **This is a key difference between the GDK and traditional Unreal networking!**

Reviewer: traditional or native?

Reviewer 6: "native" is our standard term. Also lose the "!" - we have no personality in docs (just look at me). Also, think it doesn't need to be bold.

> As the GDK works with multiple [server-workers]({{urlRoot}}/content/glossary#workers), rather than a single server, authority needs to be dictated by SpatialOS so that authority is shared between server-workers. This means server-workers have authority over some Actors but don’t have authority over other Actors, depending on how SpatialOS assigns authority. **This is a key difference between the GDK and traditional Unreal networking!**

Reviewer 3: ```suggestion
As the GDK works with multiple [server-workers]({{urlRoot}}/content/glossary#workers), rather than a single server, authority needs to be dictated by SpatialOS so that it is shared between server-workers. This means server-workers have authority over some Actors but don’t have authority over other Actors, depending on how SpatialOS assigns authority. **This is a key difference between the GDK and traditional Unreal networking!**
```
replaced repetition of "authority" with "it"

> > We use the term “authoritative” when a server-worker has authority over and Actor and “non-authoritative” when it doesn’t.

Reviewer: ```suggestion
> We use the term “authoritative” when a server-worker has authority over an Actor and “non-authoritative” when it doesn’t.
```

> > We use the term “authoritative” when a server-worker has authority over and Actor and “non-authoritative” when it doesn’t.

Reviewer 2: and-> an

> In the GDK, a server-worker is authoritative over an Actor if it has authority over the [schema]({{urlRoot}}/content/glossary#schema) [component]({{urlRoot}}/content/glossary#spatialos-component) `Position`). 

Reviewer 3: ```suggestion
In the GDK, a server-worker is authoritative over an Actor if it has authority over the [schema]({{urlRoot}}/content/glossary#schema) [component]({{urlRoot}}/content/glossary#spatialos-component) `Position`. 
```

Reviewer 5: The link here has too many brackets.

> So, in the SpatialOS GDK multiserver scenario, authority looks like this:

Reviewer: ```suggestion
So, in the SpatialOS GDK multi-server scenario, authority looks like this:
```

Actor on **authoritative server-worker**:

* `Role = ROLE_Authority`
* `RemoteRole = ROLE_SimulatedProxy`

Actor on **non-authoritative server-worker**:

* `Role = ROLE_SimulatedProxy`
* `RemoteRole = ROLE_Authority`

Actor on **[client-worker]({{urlRoot}}/content/glossary#workers)**:

* `Role = ROLE_SimulatedProxy`
* `RemoteRole = ROLE_Authority`

The GDK models owning connections using a schema component which represents the `ClientRPC`s. The worker authoritative over the `ClientRPC` schema [component]({{urlRoot}}/content/glossary#spatialos-component) is the client which owns the Actor.

In the same example as above, with the `PlayerController`, `Character` and gun, the client-worker would be authoritative over the `ClientRPC` schema component on the `PlayerController`, `Character` and gun [entities]({{urlRoot}}/content/glossary#spatialos-entity). Authority over this schema component dictates the assignment of `ROLE_AutonomousProxy`.

> Actor on **authoritative server-worker** with an **owning connection**:

Reviewer 2: Would this look better as a table?

Reviewer 3: I think all of these would, yes. I can make them if you like?

Author: 👍

* `Role = ROLE_Authority`
* `RemoteRole = ROLE_AutonomousProxy`

Actor on **non-authoritative server-worker** with an **owning connection**:

* `Role = ROLE_SimulatedProxy`
* `RemoteRole = ROLE_Authority`

Actor on **owning client-worker**:

* `Role = ROLE_AutonomousProxy`
* `RemoteRole = ROLE_Authority`

Actor on **non-owning client-worker**:

* `Role = ROLE_SimulatedProxy`
* `RemoteRole = ROLE_Authority`

> A `PlayerController` possessing different Pawns would change authority as expected; the client-worker would gain authority over the newly-possessed Pawn while losing authority over the older Pawn.

Reviewer 4: ```suggestion
A `PlayerController` possessing different Pawns would change ownership as expected; the client-worker would gain ownership over the newly-possessed Pawn while losing ownership over the older Pawn.
```
The client-worker never has authority over anything (in the Unreal sense).

> A `PlayerController` possessing different Pawns would change authority as expected; the client would gain authority over the newly-possessed Pawn while losing authority over the older Pawn.

Reviewer 3: ```suggestion
A `PlayerController` possessing different Pawns would change authority as expected; the client-worker would gain authority over the newly-possessed Pawn while losing authority over the older Pawn.
```
